src/parsec.rs

```diff
@@ -224,3 +320,100 @@ fn new_rng() -> Box<dyn Rng> {
         Box::new(rng)
     }
 }
+
+#[cfg(all(test, feature = "mock_parsec"))]
+mod tests {
+    use super::*;
+    use crate::{chain::SectionInfo, routing_table::Prefix, xor_name::XorName};
+    use unwrap::unwrap;
+
+    const DEFAULT_MIN_SECTION_SIZE: usize = 4;
+
+    #[test]
+    fn parsec_size_counter() {
+        let mut counter = ParsecSizeCounter::default();
+        assert!(!counter.needs_pruning());
+        counter.increase_size(PARSEC_SIZE_LIMIT);
+        assert!(!counter.needs_pruning());
+        counter.increase_size(1);
+        assert!(counter.needs_pruning());
+    }
+
+    fn create_full_ids() -> Vec<FullId> {
+        (0..DEFAULT_MIN_SECTION_SIZE)
+            .map(|_| FullId::new())
+            .collect()
+    }
+
+    fn create_gen_pfx_info(full_ids: Vec<FullId>, version: u64) -> GenesisPfxInfo {
+        let members = full_ids.iter().map(|id| *id.public_id()).collect();
+        let section_info = unwrap!(SectionInfo::new_for_test(
+            members,
+            Prefix::<XorName>::default(),
+            version
+        ));
+        GenesisPfxInfo {
+            first_info: section_info,
+            first_state_serialized: Vec::new(),
+            latest_info: SectionInfo::default(),
+        }
+    }
+
+    // Create a ParsecMap with 3 sections
+    fn create_parsec_map() -> ParsecMap {
+        let log_ident = LogIdent::new("node");
+        let full_ids = create_full_ids();
+        let full_id = full_ids[0].clone();
+
+        let gen_pfx_info = create_gen_pfx_info(full_ids.clone(), 0);
+        let mut parsec_map = ParsecMap::new(full_id.clone(), &gen_pfx_info);
+
+        let gen_pfx_info = create_gen_pfx_info(full_ids.clone(), 1);
+        parsec_map.init(full_id.clone(), &gen_pfx_info, &log_ident);
+
+        let gen_pfx_info = create_gen_pfx_info(full_ids, 2);
+        parsec_map.init(full_id, &gen_pfx_info, &log_ident);
+
+        parsec_map
+    }
+
+    fn parsec_map_to_vec(map: BTreeMap<u64, ParsecAndCounter>) -> Vec<(u64, ParsecSizeCounter)> {
+        map.into_iter()
+            .map(|(id, ParsecAndCounter { size_counter, .. })| (id, size_counter))
+            .collect()
+    }
+
+    #[test]
+    fn parsec_map_check_prune_required() {
+        let response = Response::new();
+        let full_id = FullId::new();
+        let pub_id = full_id.public_id();
+
+        let log_ident = LogIdent::new("node");
+        let mut parsec_map = create_parsec_map();
+
+        // First handle responses with parsec 1 which should not trigger a need for pruning.
+        for _ in 0..=PARSEC_SIZE_LIMIT {
+            assert!(!parsec_map.handle_response(1, response.clone(), *pub_id, &log_ident));
```

Comment:
Why do we assert on the handle_response result? Is it not assume it will be true for this test?

Reply:
Just to assert that we're not sending to the latest parsec

Reply:
Completely missed that bit can you clarify that as well.
Do we really care about it? I mean if we do send it the the latest parsec we should get a failure because need_pruning would be false.

Reply:
Yes the assert afterwards would also fail (but not necessarily for that reason if something else is also broken). Why would we not check the return value for `handle_response`?

Reply:
The reason, is that is probably a different unit test, i.e it is a handle_response test. As a result it is less clear what you are actually testing in this code. The AAA pattern is a guide that allow to clarify test, the assert should not be used to validate an assumption in the test, but what we want to validate.

So what you could do is add a small test that call handle_response on the different parsec and validate it is correct, and then just assume in this test it is.

Reply:
Well, the assert here isn't testing that `handle_response` returns false when sending to and old parsec, it's verifying for the purpose of the later asserts that we didn't send to the latest parsec

Reply:
If that is true, would not not be better done declaratively?
(i.e there is an implicit assumption here that last parsec means return true, and that is tested nowhere).

```
const INITIAL_PARSEC = 0;
const FIRST_PARSEC = 1;
const LAST_PARSEC = 2;

handle_n_parsec_responses(PARSEC_SIZE_LIMIT, FIRST_PARSEC, ... );
```

Reply:
I like having these named constants, so I'll add that. But having a general function like this still doesn't confirm that FIRST_PARSEC isn't handled as the last parsec

Reply:
No, but you have to make choices on what your test do validate otherwise very quickly become difficult to read.

You want to validate that assumption make it the other test I suggested (i.e test handle_parsec_response(FIRST_PARSEC ... ) is false and LAST_PARSEC is true.

What I am saying is you are testing multiple things in your test. That is why guildline like the AAA pattern where created, to help keep test clean.

Now it is up to you to decide.

Reply:
Even in a strict AAA pattern, checking that the steps in the ACT phase are doing what you think it's doing makes things easier I think. The earlier the failure happens the easier it's to spot what's wrong.

That said, following this discussion I've split the test into 2 to follow AAA, and added a third test to test requests too